cpp/src/knn/knn.cu

```diff
@@ -142,40 +153,22 @@ void kNN::fit(float **input, int *sizes, int N) {
 	 * @param res_I 	   pointer to device memory for returning k nearest indices
 	 * @param res_D		   pointer to device memory for returning k nearest distances
 	 * @param k			   number of neighbors to query
+	 * @param rowMajor is the query array in row major layout?
 	 */
-void kNN::search(float *search_items, int n, long *res_I, float *res_D, int k) {
-  MLCommon::Selection::brute_force_knn(ptrs, sizes, indices, D, search_items, n,
-                                       res_I, res_D, k,
-                                       handle->getImpl().getStream());
-}
+void kNN::search(float *search_items, int n, int64_t *res_I, float *res_D,
+                 int k, bool rowMajor) {
+  ASSERT(this->indices > 0, "Cannot search before model has been trained.");
 
-/**
-     * Chunk a host array up into one or many GPUs (determined by the provided
-     * list of gpu ids) and fit a knn model.
-     *
-     * @param ptr       an array in host memory to chunk over devices
-     * @param n         number of elements in ptr
-     * @param devices   array of device ids for chunking the ptr
-     * @param n_chunks  number of elements in gpus
-     * @param out       host pointer (size n) to store output
-     */
-void kNN::fit_from_host(float *ptr, int n, int *devices, int n_chunks) {
-  if (this->owner)
-    for (int i = 0; i < this->indices; i++) {
-      CUDA_CHECK(cudaFree(this->ptrs[i]));
-    }
-
-  reset();
-
-  this->owner = true;
-
-  float **params = new float *[n_chunks];
-  int *sizes = new int[n_chunks];
-
-  chunk_to_device<float>(ptr, n, D, devices, params, sizes, n_chunks,
-                         handle->getImpl().getStream());
+  cudaStream_t int_streams[handle->getImpl().getNumInternalStreams()];
+  for (int i = 0; i < handle->getImpl().getNumInternalStreams(); i++) {
+    int_streams[i] = handle->getImpl().getInternalStream(i);
+  }
```

Comment:
Since this seems to be a repeating pattern. How about putting this inside `cumlHandle_impl` itself?